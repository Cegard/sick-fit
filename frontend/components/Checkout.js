import { useMutation } from '@apollo/client';
import {
  CardElement,
  Elements,
  useElements,
  useStripe,
} from '@stripe/react-stripe-js';
import { loadStripe } from '@stripe/stripe-js';
import gql from 'graphql-tag';
import { useRouter } from 'next/dist/client/router';
import nProgress from 'nprogress';
import { useState } from 'react';
import styled from 'styled-components';
import { useCart } from '../lib/cartState';
import DisplayError from './ErrorMessage';
import SickButton from './styles/SickButton';
import { CURRENT_USER_QUERY } from './User';

const CheckoutFormStyles = styled.form`
  box-shadow: 0 1px 2px 2px rgba(0, 0, 0, 0.04);
  border: 1px solid rgba(0, 0, 0, 0.06);
  border-radius: 5px;
  display: grid;
  grid-gap: 1rem;
`;

const CREATE_ORDER_MUTATION = gql`
  mutation CREATE_ORDER_MUTATION($token: String!) {
    checkout(token: $token) {
      id
    }
  }
`;

const stripeLib = loadStripe(process.env.NEXT_PUBLIC_STRIPE_KEY);

function CheckoutForm() {
  const [error, setError] = useState();
  const [loading, setLoading] = useState(false);
  const stripe = useStripe();
  const elements = useElements();
  const router = useRouter();
  const { closeCart } = useCart();
  const [checkout, { loading: loadingCheckout, error: errorDuringCheckout }] =
    useMutation(CREATE_ORDER_MUTATION, {
      refetchQueries: [{ query: CURRENT_USER_QUERY }],
    });

  async function handleSubmit(e) {
    e.preventDefault();
    setLoading(true);
    nProgress.start();
    const { error: errorDuringPayment, paymentMethod } =
      await stripe.createPaymentMethod({
        type: 'card',
        card: elements.getElement(CardElement),
      });

    if (errorDuringPayment) {
      setError(errorDuringPayment);
      setLoading(false);
      nProgress.done();
      return;
    }

    const order = await checkout({ variables: { token: paymentMethod.id } });

    setLoading(false);
    nProgress.done();
    closeCart();
    router.push({
      pathname: '/orders',
      query: { id: order.data.checkout.id },
    });
  }

  return (
    <CheckoutFormStyles onSubmit={(e) => handleSubmit(e)}>
      {error && <DisplayError error={error} />}
      <CardElement />
      <SickButton>Check Out Now</SickButton>
    </CheckoutFormStyles>
  );
}

export default function Checkout() {
  return (
    <Elements stripe={stripeLib}>
      <CheckoutForm />
    </Elements>
  );
}
