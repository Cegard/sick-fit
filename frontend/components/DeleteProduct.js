import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';

const MUTATION_DELETE_PRODUCT = gql`
  mutation MUTATION_DELETE_PRODUCT($id: ID!) {
    deleteProduct(id: $id) {
      id
      name
    }
  }
`;

function update(cache, payload) {
  cache.evict(cache.identify(payload.data.deleteProduct));
}

export default function DeleteProduct({ id, children }) {
  const [deleteProduct, { loading }] = useMutation(MUTATION_DELETE_PRODUCT, {
    variables: {
      id,
    },
    update,
  });

  return (
    <button
      type="button"
      disabled={loading}
      onClick={() => {
        if (confirm('Are you sure?')) {
          deleteProduct().catch((err) => alert(err.message));
        }
      }}
    >
      {children}
    </button>
  );
}
