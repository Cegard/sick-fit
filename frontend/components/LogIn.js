import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import Router from 'next/router';
import Form from './styles/Form';
import useForm from '../lib/useForm';
import { CURRENT_USER_QUERY } from './User';
import DisplayError from './ErrorMessage';

const LOGIN_MUTATION = gql`
  mutation LOGIN_MUTATION($email: String!, $password: String!) {
    authenticateUserWithPassword(email: $email, password: $password) {
      ... on UserAuthenticationWithPasswordSuccess {
        item {
          id
          email
          name
        }
      }
      ... on UserAuthenticationWithPasswordFailure {
        code
        message
      }
    }
  }
`;

export default function LogIn() {
  const { inputs, handleChange, resetForm } = useForm({
    email: '',
    password: '',
  });
  const [login, { data, loading }] = useMutation(LOGIN_MUTATION, {
    variables: { ...inputs },
    refetchQueries: [{ query: CURRENT_USER_QUERY }],
  });

  async function handleSubmit(e) {
    e.preventDefault();
    const res = await login();
    console.log({ data, res });
    resetForm();

    if (
      res.data.authenticateUserWithPassword.__typename ===
      'UserAuthenticationWithPasswordSuccess'
    ) {
      Router.push({
        pathname: `/products`,
      });
    }
  }

  const error =
    data?.authenticateUserWithPassword.__typename ===
    'UserAuthenticationWithPasswordFailure'
      ? data.authenticateUserWithPassword
      : null;

  return (
    <Form method="POST" onSubmit={(e) => handleSubmit(e)}>
      <h2> Log In! </h2>
      <DisplayError error={error} />
      <fieldset>
        <label htmlFor="email">
          Email
          <input
            type="email"
            name="email"
            placeholder="email@example.com"
            autoComplete="email"
            value={inputs.email}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="password">
          Password
          <input
            type="password"
            name="password"
            placeholder="your password"
            value={inputs.password}
            onChange={handleChange}
          />
        </label>
        <button type="submit"> Log In </button>
      </fieldset>
    </Form>
  );
}
