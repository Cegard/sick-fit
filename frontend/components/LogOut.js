import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import Router from 'next/router';
import { CURRENT_USER_QUERY } from './User';

const LOGOUT_MUTATION = gql`
  mutation {
    endSession
  }
`;

export default function LogOut() {
  const [logout] = useMutation(LOGOUT_MUTATION, {
    refetchQueries: [{ query: CURRENT_USER_QUERY }],
  });

  async function handleLogOut() {
    await logout();
    Router.push({
      pathname: '/',
    });
  }

  return (
    <>
      <button type="button" onClick={() => handleLogOut()}>
        Log out
      </button>
    </>
  );
}
