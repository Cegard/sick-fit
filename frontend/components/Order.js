import { useQuery } from '@apollo/client';
import gql from 'graphql-tag';
import Head from 'next/head';
import DisplayError from './ErrorMessage';
import OrderStyles from './styles/OrderStyles';
import formatMoney from '../lib/formatMoney';

const ORDER_QUERY = gql`
  query ORDER_QUERY($id: ID!) {
    order: Order(where: { id: $id }) {
      id
      charge
      total
      user {
        id
      }
      items {
        id
        name
        description
        price
        quantity
        image {
          id
          image {
            id
            publicUrlTransformed
          }
        }
      }
    }
  }
`;

export default function Order({ id }) {
  const { data, loading, error } = useQuery(ORDER_QUERY, {
    variables: { id },
  });

  if (loading) return <p>Loading...</p>;

  if (error) return <DisplayError error={error} />;

  const { order } = data;

  return (
    <OrderStyles>
      <Head>
        <title>Sick Fits - {order.id}</title>
      </Head>
      <div>
        <p>
          <span>Order Id: </span>
          <span>{id}</span>
        </p>
        <p>
          <span>Charge: </span>
          <span>{order.charge}</span>
        </p>
        <p>
          <span>Total: </span>
          <span>{formatMoney(order.total)}</span>
        </p>
      </div>
      <div className="items">
        {order.items.map((item) => (
          <div className="order-item" key={item.id}>
            <img src={item.image.image.publicUrlTransformed} alt={item.name} />
            <h2>{item.name}</h2>
            <p>Quantity: {item.quantity}</p>
            <p>Each: {formatMoney(item.price)}</p>
            <p>Sub Total: {formatMoney(item.quantity * item.price)}</p>
            <p>{item.description}</p>
          </div>
        ))}
      </div>
    </OrderStyles>
  );
}
