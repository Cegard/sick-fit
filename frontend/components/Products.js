import { useQuery } from '@apollo/client';
import gql from 'graphql-tag';
import styled from 'styled-components';
import Product from './Product';
import { perPage } from '../config';

export const ALL_PRODUCTS_QUERY = gql`
  query ALL_PRODUCTS_QUERY($first: Int, $skip: Int = 0) {
    allProducts(first: $first, skip: $skip) {
      id
      name
      price
      description
      image {
        id
        altText
        image {
          publicUrlTransformed
        }
      }
    }
  }
`;

const StyledProductsList = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-gap: 60px;
`;

export default function ProductPage({ page }) {
  const pageNumber = parseInt(page);
  const { data, err, loading } = useQuery(ALL_PRODUCTS_QUERY, {
    variables: {
      skip: pageNumber * perPage - perPage,
      first: perPage,
    },
  });

  if (loading) return <p>Loading...</p>;

  if (err) return <p>Error: {err.message}</p>;

  return (
    <StyledProductsList>
      {data &&
        data.allProducts.map((product) => (
          <Product key={product.name} product={product} />
        ))}
    </StyledProductsList>
  );
}
