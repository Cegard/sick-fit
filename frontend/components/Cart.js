import styled from 'styled-components';
import { useUser } from './User';
import CartStyles from './styles/CartStyles';
import Supreme from './styles/Supreme';
import formatMoney from '../lib/formatMoney';
import { useCart } from '../lib/cartState';
import CloseButton from './styles/CloseButton';
import RemoveFromCart from './RemoveFromCart';
import Checkout from './Checkout';

const StyledLi = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid var(--lighGrey);
  display: grid;
  grid-template-columns: auto 1fr auto;
  img {
    margin-right: 1rem;
  }
  h3,
  p {
    margin: 0;
  }
`;

function CartItem({ cartItem }) {
  const { product } = cartItem;

  if (!product) return <></>;

  return (
    <StyledLi>
      <img
        width="100"
        src={product.image.image.publicUrlTransformed}
        alt={product.name}
      />
      <div>
        <h3>{product.name}</h3>
        <p>
          {formatMoney(product.price * cartItem.quantity)}-
          <em>
            {cartItem.quantity} &times; {formatMoney(product.price)} each
          </em>{' '}
        </p>
      </div>
      <RemoveFromCart productId={cartItem.id} />
    </StyledLi>
  );
}

export default function Cart() {
  const user = useUser();
  const { cartOpen, closeCart } = useCart();

  if (!user || !user?.cart) return <></>;

  return (
    <CartStyles open={cartOpen}>
      <header>
        <Supreme>{user.name}'s cart</Supreme>
        <CloseButton onClick={closeCart}>X</CloseButton>
      </header>
      <ul>
        {user.cart.map((item) => (
          <CartItem key={item.id} cartItem={item} />
        ))}
      </ul>
      <footer>
        <div>
          {formatMoney(
            user.cart.reduce((accounted, currItem) => {
              if (!currItem?.product) return accounted;

              return accounted + currItem.product.price * currItem.quantity;
            }, 0)
          )}
        </div>
        <div>
          <Checkout />
        </div>
      </footer>
    </CartStyles>
  );
}
