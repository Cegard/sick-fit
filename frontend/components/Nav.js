import Link from 'next/link';
import NavStyles from './styles/NavStyles';
import { useUser } from './User';
import LogOut from './LogOut';
import { useCart } from '../lib/cartState';
import CartCount from './CartCount';

export default function Nav() {
  const user = useUser();
  const { openCart } = useCart();

  return (
    <NavStyles>
      <Link href="/products"> Products </Link>
      {user && (
        <>
          <Link href="/orders"> Orders </Link>
          <Link href="/sell"> Sell </Link>
          <Link href="/account"> Account </Link>
          <LogOut />
          <button type="button" onClick={openCart}>
            My Cart
            <CartCount
              count={user.cart.reduce(
                (accounted, currItem) => accounted + currItem.quantity,
                0
              )}
            />
          </button>
        </>
      )}
      {!user && <Link href="/login"> Log in </Link>}
    </NavStyles>
  );
}
