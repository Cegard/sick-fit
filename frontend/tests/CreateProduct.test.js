import { render, screen, waitFor } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import userEvent from '@testing-library/user-event';
import Router from 'next/router';
import wait from 'waait';
import CreateProduct, {
  CREATE_PRODUCT_MUTATION,
} from '../components/CreateProduct';
import { fakeItem } from '../lib/testUtils';
import { ALL_PRODUCTS_QUERY } from '../components/Products';

const fakedItem = fakeItem();
jest.mock('next/router', () => ({
  push: jest.fn(),
}));

describe('<CreateProduct/>', () => {
  it('Renders and matches snapshot', () => {
    const { container } = render(
      <MockedProvider>
        <CreateProduct />
      </MockedProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it('Handles the updating', async () => {
    const { container } = render(
      <MockedProvider>
        <CreateProduct />
      </MockedProvider>
    );

    const productNameInput = screen.getByPlaceholderText(/Name/);
    const productPriceInput = screen.getByPlaceholderText('0');
    const productDescriptionInput = screen.getByPlaceholderText('');

    await userEvent.type(productNameInput, fakedItem.name);
    await userEvent.type(productPriceInput, fakedItem.price.toString());
    await userEvent.type(productDescriptionInput, fakedItem.description);

    const populatedProductName = screen.getByDisplayValue(fakedItem.name);
    const populatedProductPrice = screen.getByDisplayValue(fakedItem.price);
    const populatedProductDescription = screen.getByDisplayValue(
      fakedItem.description
    );

    expect(populatedProductName).toBeInTheDocument();
    expect(populatedProductPrice).toBeInTheDocument();
    expect(populatedProductDescription).toBeInTheDocument();
  });

  it('Creates the items when the form is submitted', async () => {
    const mocks = [
      {
        request: {
          query: CREATE_PRODUCT_MUTATION,
          variables: {
            name: fakedItem.name,
            description: fakedItem.description,
            price: fakedItem.price,
            image: '',
          },
        },
        result: {
          data: {
            createProduct: { ...fakedItem, id: 'abc123', __typename: 'Item' },
          },
        },
      },
      {
        request: {
          query: ALL_PRODUCTS_QUERY,
          variables: { skip: 0, first: 2 },
        },
        result: { data: { allProducts: [fakedItem] } },
      },
    ];

    const { container } = render(
      <MockedProvider mocks={mocks}>
        <CreateProduct />
      </MockedProvider>
    );

    const productNameInput = screen.getByPlaceholderText(/Name/);
    const productPriceInput = screen.getByPlaceholderText('0');
    const productDescriptionInput = screen.getByPlaceholderText('');
    const submitButton = screen.getByText('+ Add Product');

    await userEvent.type(productNameInput, fakedItem.name);
    await userEvent.type(productPriceInput, fakedItem.price.toString());
    await userEvent.type(productDescriptionInput, fakedItem.description);
    await userEvent.click(submitButton);

    await waitFor(() => wait(0));

    expect(Router.push).toHaveBeenCalled();
    expect(Router.push).toHaveBeenCalledWith({ pathname: '/product/abc123' });
  });
});
