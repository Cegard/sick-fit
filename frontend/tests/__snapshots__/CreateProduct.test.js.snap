// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateProduct/> Renders and matches snapshot 1`] = `
<div>
  <form
    class="Form-sc-1xszr8q-0 iaxOhV"
  >
    <fieldset
      aria-busy="false"
    >
      <label
        for="image"
      >
        Image
        <input
          id="image"
          name="image"
          type="file"
        />
      </label>
      <label
        for="name"
      >
        Name
        <input
          id="name"
          name="name"
          placeholder="Name"
          required=""
          type="text"
          value=""
        />
      </label>
      <label
        for="price"
      >
        Price
        <input
          id="price"
          name="price"
          placeholder="0"
          required=""
          type="number"
          value="0"
        />
      </label>
      <label
        for="description"
      >
        Description
        <textarea
          id="description"
          name="description"
          placeholder=""
          required=""
        />
      </label>
      <button
        type="submit"
      >
        + Add Product
      </button>
    </fieldset>
  </form>
</div>
`;
